lib/elixir/lib/application.ex

```diff
@@ -8,78 +8,100 @@ defmodule Application do
-
-  The environment is available after loading the application, which is a process
-  explained later:
+  The environment is available after loading the application. This is typically
+  done automatically for you by tools like Mix, but it can also be done explicitly
+  like so:
 
       Application.load(:APP_NAME)
       #=> :ok
 
-      Application.get_env(:APP_NAME, :redis_host)
+      Application.fetch_env!(:APP_NAME, :redis_host)
       #=> "localhost"
 
   In Mix projects, the environment of the application and its dependencies can
-  be overridden via the `config/config.exs` file. If you start the application
-  with Mix, that configuration is available at compile time, and at runtime too,
-  but take into account it is not included in the generated application resource
-  file, and it is not available if you start the application without Mix.
-
-  For example, someone using your application can override its `:redis_host`
-  environment variable as follows:
+  be overridden via the `config/config.exs` file. For example, someone using
+  your application can override its `:redis_host` environment variable as follows:
 
-      config :APP_NAME, redis_host: "redis.local"
+      import Config
+      config :APP_NAME, :redis_host, "redis.local"
 
   The function `put_env/3` allows dynamic configuration of the application
   environment, but as a rule of thumb each application is responsible for its
   own environment. Please do not use the functions in this module for directly
   accessing or modifying the environment of other applications.
 
-  The application environment can be overridden via the `-config` option of
-  `erl`, as well as command-line options, as we are going to see below.
+  ### Compile-time environment
+
+  The application environment is available both at compile time and at runtime.
+  For example, you can use the application environment and the configuration
+  system to specify a value that is read when the code is being compiled:
+
+      config :APP_NAME, :redis_host, "redis.local"
+
+  And then in the code:
+
+      defmodule RedisClient do
+        @host Application.fetch_env!(:APP_NAME, :redis_host)
+
+        def connect do
+          # access @host here
+        end
+      end
+
+  This approach, however, has one big limitation: if the Redis host is set
+  dynamically, for example through `System.get_env/2` or only when booting
+  the application, the code won't be recompiled. This can lead into situations
+  where the configuration the code was compiled with is different from the
+  code running, which can be cumbersome to debug.
+
+  Therefore, whenever reading the environment at compilation time, we
+  recommend using `Application.compile_env!/2` instead:
+
+      @host Application.compile_env!(:APP_NAME, :redis_host)
+
+  When `compile_env!/2` is used, Elixir is capable of notifying whenever
+  those values differ between compilation time and runtime.
```

Comment:
This is where some expansion could be useful. How exactly is Elixir capable of notifying, i.e. what happens when those values differ?